File: server/internal/adapter/handler/dto.go
package handler

import (
	"time"

	"github.com/google/uuid"

	"github.com/kotarowakuwaku/asset-wish/server/internal/domain"
	"github.com/kotarowakuwaku/asset-wish/server/internal/usecase"
)

// レスポンスのキーは lowerCamelCase、金額は整数（円）、日付は YYYY-MM-DD
// （detailed-design 6）。金額を文字列にしないのは、クライアント側で
// 数値として扱うため。JavaScript の安全な整数の範囲に十分収まる。

// dateLayout は日付のみの表現。
const dateLayout = "2006-01-02"

type accountResponse struct {
	ID        uuid.UUID `json:"id"`
	Name      string    `json:"name"`
	Kind      string    `json:"kind"`
	Balance   int64     `json:"balance"`
	UpdatedAt time.Time `json:"updatedAt"`
	// IsStale は残高の更新を促すための導出値。
	IsStale bool `json:"isStale"`
}

func newAccountResponse(a domain.Account, now time.Time) accountResponse {
	return accountResponse{
		ID:        a.ID,
		Name:      a.Name,
		Kind:      string(a.Kind),
		Balance:   int64(a.Balance),
		UpdatedAt: a.UpdatedAt,
		IsStale:   a.IsStale(now, domain.StaleBalanceThreshold),
	}
}

type lendingResponse struct {
	ID           uuid.UUID `json:"id"`
	Counterparty string    `json:"counterparty"`
	Description  string    `json:"description"`
	Amount       int64     `json:"amount"`
	// CollectedAmount / Outstanding / Status のうち、DB が持つのは
	// CollectedAmount だけ。残りは導出値（不変条件12）。
	CollectedAmount int64  `json:"collectedAmount"`
	Outstanding     int64  `json:"outstanding"`
	Status          string `json:"status"`
	OccurredOn      string `json:"occurredOn"`
}

func newLendingResponse(l domain.Lending) lendingResponse {
	return lendingResponse{
		ID:              l.ID,
		Counterparty:    l.Counterparty,
		Description:     l.Description,
		Amount:          int64(l.Amount),
		CollectedAmount: int64(l.CollectedAmount),
		Outstanding:     int64(l.Outstanding()),
		Status:          string(l.Status()),
		OccurredOn:      l.OccurredOn.Format(dateLayout),
	}
}

type wishResponse struct {
	ID       uuid.UUID `json:"id"`
	Title    string    `json:"title"`
	Amount   int64     `json:"amount"`
	Category string    `json:"category"`
	Status   string    `json:"status"`
	Priority int       `json:"priority"`
	// Deadline は未設定なら null。
	Deadline *string `json:"deadline"`
}

func newWishResponse(w domain.Wish) wishResponse {
	return wishResponse{
		ID:       w.ID,
		Title:    w.Title,
		Amount:   int64(w.Amount),
		Category: string(w.Category),
		Status:   string(w.Status),
		Priority: w.Priority,
		Deadline: formatDatePtr(w.Deadline),
	}
}

type monthlyBalanceResponse struct {
	ID        uuid.UUID `json:"id"`
	YearMonth string    `json:"yearMonth"`
	Income    int64     `json:"income"`
	Expense   int64     `json:"expense"`
	// Surplus は導出値。クライアントは符号で黒字・赤字を判定する。
	Surplus int64 `json:"surplus"`
}

func newMonthlyBalanceResponse(m domain.MonthlyBalance) monthlyBalanceResponse {
	return monthlyBalanceResponse{
		ID:        m.ID,
		YearMonth: m.YearMonth.String(),
		Income:    int64(m.Income),
		Expense:   int64(m.Expense),
		Surplus:   int64(m.Surplus()),
	}
}

type transactionResponse struct {
	ID         uuid.UUID  `json:"id"`
	AccountID  uuid.UUID  `json:"accountId"`
	Amount     int64      `json:"amount"`
	Kind       string     `json:"kind"`
	RefID      *uuid.UUID `json:"refId"`
	OccurredOn string     `json:"occurredOn"`
}

func newTransactionResponse(t domain.Transaction) transactionResponse {
	return transactionResponse{
		ID:         t.ID,
		AccountID:  t.AccountID,
		Amount:     int64(t.Amount),
		Kind:       string(t.Kind),
		RefID:      t.RefID,
		OccurredOn: t.OccurredOn.Format(dateLayout),
	}
}

type dashboardResponse struct {
	NetAsset        int64             `json:"netAsset"`
	Breakdown       breakdownResponse `json:"breakdown"`
	InvestmentTotal int64             `json:"investmentTotal"`
	// AverageSurplus は HasAverageSurplus が false のとき 0 を返すが、
	// クライアントは表示しない（detailed-design 6.1）。
	AverageSurplus    int64              `json:"averageSurplus"`
	HasAverageSurplus bool               `json:"hasAverageSurplus"`
	Wishes            []dashboardWishRes `json:"wishes"`
}

type breakdownResponse struct {
	CashTotal           int64 `json:"cashTotal"`
	OutstandingLendings int64 `json:"outstandingLendings"`
	Commitments         int64 `json:"commitments"`
}

type dashboardWishRes struct {
	wishResponse
	Shortfall int64 `json:"shortfall"`
	// MonthsToReach は算出不可なら null。クライアントは null を
	// 「算出不可」と表示する。0 を返すと「今月中に届く」と誤読される。
	MonthsToReach *int `json:"monthsToReach"`
}

func newDashboardResponse(d usecase.Dashboard) dashboardResponse {
	wishes := make([]dashboardWishRes, 0, len(d.Wishes))
	for _, w := range d.Wishes {
		var months *int
		if w.HasMonthsToReach {
			m := w.MonthsToReach
			months = &m
		}
		wishes = append(wishes, dashboardWishRes{
			wishResponse:  newWishResponse(w.Wish),
			Shortfall:     int64(w.Shortfall),
			MonthsToReach: months,
		})
	}

	return dashboardResponse{
		NetAsset: int64(d.NetAsset),
		Breakdown: breakdownResponse{
			CashTotal:           int64(d.Breakdown.CashTotal),
			OutstandingLendings: int64(d.Breakdown.OutstandingLendings),
			Commitments:         int64(d.Breakdown.Commitments),
		},
		InvestmentTotal:   int64(d.InvestmentTotal),
		AverageSurplus:    int64(d.AverageSurplus),
		HasAverageSurplus: d.HasAverageSurplus,
		Wishes:            wishes,
	}
}

func formatDatePtr(t *time.Time) *string {
	if t == nil {
		return nil
	}
	s := t.Format(dateLayout)
	return &s
}

// parseDate は YYYY-MM-DD を UTC の 00:00 として解釈する。
// タイムゾーンを持ち込むと、DB の DATE 列との往復で日がずれる。
func parseDate(s string) (time.Time, error) {
	t, err := time.Parse(dateLayout, s)
	if err != nil {
		return time.Time{}, newBadRequest("INVALID_DATE", "日付は YYYY-MM-DD 形式で指定してください")
	}
	return t, nil
}

File: worker/src/adapter/handler/dto.ts
// レスポンスのキーは lowerCamelCase、金額は整数（円）、日付は YYYY-MM-DD。
// 金額を文字列にしないのは、クライアント側で数値として扱うため。
// JavaScript の安全な整数の範囲に十分収まる。
//
// **front/src/api/types.ts と一対一で対応させる。** 片方だけ変えると、
// 型は通るのに実行時に undefined を触ることになる。

import { STALE_BALANCE_THRESHOLD_MS, type Account } from '../../domain/account'
import type { Loan } from '../../domain/loan'
import type { MonthlySummary } from '../../domain/monthlySummary'
import type { Instant } from '../../domain/time'
import type { Transaction } from '../../domain/transaction'
import type { Wish } from '../../domain/wish'
import type { Dashboard } from '../../usecase/dashboard'

export function accountResponse(a: Account, now: Instant) {
  return {
    id: a.id,
    name: a.name,
    kind: a.kind,
    balance: a.balance,
    updatedAt: a.updatedAt,
    // 残高の更新を促すための導出値。
    isStale: a.isStale(now, STALE_BALANCE_THRESHOLD_MS),
  }
}

export function loanResponse(l: Loan) {
  return {
    id: l.id,
    // 'lent' か 'borrowed'。金額は向きによらず正なので、向きはこれだけが表す。
    direction: l.direction,
    counterparty: l.counterparty,
    description: l.description,
    amount: l.amount,
    // settledAmount 以外は導出値。DB は持たない（不変条件12）。
    settledAmount: l.settledAmount,
    outstanding: l.outstanding(),
    status: l.status(),
    occurredOn: l.occurredOn,
  }
}

export function wishResponse(w: Wish) {
  return {
    id: w.id,
    title: w.title,
    amount: w.amount,
    category: w.category,
    status: w.status,
    priority: w.priority,
    // 未設定なら null。
    deadline: w.deadline,
  }
}

/**
 * 月次の集計。
 *
 * **id を持たない。** 明細から導出した値であって、保存された行ではない。
 * id を載せると、更新できる資源に見えてしまう。
 */
export function monthlySummaryResponse(s: MonthlySummary) {
  return {
    yearMonth: s.yearMonth.toString(),
    income: s.income,
    expense: s.expense,
    // 導出値。クライアントは符号で黒字・赤字を判定する。
    surplus: s.surplus(),
    // 'entries' なら明細の集計、'manual' なら明細が無い月を手入力で埋めたもの。
    // どちらを見ているか分からないと、「明細を打ったのに反映されない」ように見える。
    source: s.source,
  }
}

export function transactionResponse(t: Transaction) {
  return {
    id: t.id,
    accountId: t.accountId,
    amount: t.amount,
    kind: t.kind,
    refId: t.refId,
    occurredOn: t.occurredOn,
    // 手入力の明細以外では空。何の取引かは refId をたどれば分かる。
    note: t.note,
  }
}

export function dashboardResponse(d: Dashboard) {
  return {
    netAsset: d.netAsset,
    breakdown: {
      cashTotal: d.breakdown.cashTotal,
      commitments: d.breakdown.commitments,
    },
    // investmentTotal と outstanding* はどちらも実質資産の外の参考値。
    // breakdown に混ぜないのは、合計に足されない値だと形で示すため
    // （不変条件1・4）。
    investmentTotal: d.investmentTotal,
    // 貸しと借りは分けて返す。差額にすると、誰にいくら貸しているのかが消える。
    outstandingLent: d.outstanding.lent,
    outstandingBorrowed: d.outstanding.borrowed,
    // 算出不可のときは 0 を返すが、hasAverageSurplus が false なので
    // クライアントは表示しない。
    averageSurplus: d.averageSurplus ?? 0,
    hasAverageSurplus: d.averageSurplus !== null,
    wishes: d.wishes.map((w) => ({
      ...wishResponse(w.wish),
      shortfall: w.shortfall,
      // 算出不可なら null。0 を返すと「今月中に届く」と誤読される。
      monthsToReach: w.monthsToReach,
      // 期限までに毎月いくら貯めればよいか。期限が無ければ null。
      monthlySavingNeeded: w.monthlySavingNeeded,
    })),
  }
}
